### PR TITLE
Exclude label setting panel

### PR DESCRIPTION
**Pull request by amit-shar**
tests have been written for exclude custom imports

### CODE REVIEW

File: plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/SettingsPanel.java [hunk 1]
```diff
@@ -97,49 +102,66 @@ public final JComponent createComponent() {
...
+
+            return jPanel;
+        }
 
-    @Override
-    public final boolean isModified() {
-        return true;
-    }
+        @Override
+        public final boolean isModified() {
```
Comment (ScrapCodes) [inline plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/SettingsPanel.java]: Can you fix your formatting

File: plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/EditorDocOps.java [hunk 2]
```diff
@@ -159,4 +182,44 @@ private void getCompletePackage(final PsiPackage completePackage,
         }
         return externalImports;
     }
+
+    public final Set<String> excludeCustomImports(final Set<String> imports,
```
Comment (ScrapCodes) [inline plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/EditorDocOps.java]: rename -> excludeConfiguredImports

File: plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/SettingsPanel.java [hunk 3]
```diff
@@ -97,46 +102,63 @@ public final JComponent createComponent() {
...
+        excludeImportsText.setEditable(true);
+        excludeImportsText.setVisible(true);
 
-        return jPanel;
-    }
+        if (propertiesComponent.isValueSet(RefreshAction.EXCLUDE_IMPORT_LIST)) {
+            excludeImportsText.setText(propertiesComponent.getValue(RefreshAction.
+                    EXCLUDE_IMPORT_LIST));
+        }
+            JPanel jPanel = new JPanel(formLayout);
```
Comment (ScrapCodes) [inline plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/SettingsPanel.java]: Fix the indentation on these lines.

File: plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/SettingsPanel.java [hunk 4]
```diff
@@ -97,46 +102,63 @@ public final JComponent createComponent() {
...
-        PropertiesComponent propertiesComponent = PropertiesComponent.getInstance();
-        String esURLValue = esURLText.getText();
-        String sizeValue = sizeText.getText();
-        String distanceValue = distanceText.getText();
-        propertiesComponent.setValue(RefreshAction.ES_URL, esURLValue);
-        propertiesComponent.setValue(RefreshAction.SIZE, sizeValue);
-        propertiesComponent.setValue(RefreshAction.DISTANCE, distanceValue);
-    }
+        @Override
+        public final void apply() {
```
Comment (ScrapCodes) [inline plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/SettingsPanel.java]: Fix indentation here.

File: plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/SettingsPanel.java [hunk 5]
```diff
@@ -97,46 +102,63 @@ public final JComponent createComponent() {
...
-                                    RefreshAction.ES_URL_DEFAULT));
-        sizeText.setText(propertiesComponent.
-                            getValue(RefreshAction.SIZE,
-                                    String.valueOf(RefreshAction.SIZE_DEFAULT_VALUE)));
-        distanceText.setText(propertiesComponent.
-                            getValue(RefreshAction.DISTANCE,
-                                    String.valueOf(RefreshAction.DISTANCE_DEFAULT_VALUE)));
-    }
+        @Override
+        public final void reset() {
```
Comment (ScrapCodes) [inline plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/SettingsPanel.java]: here as well.

File: plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/EditorDocOps.java [hunk 6]
```diff
@@ -85,17 +93,31 @@
...
-            String line = document.getCharsSequence().
-                                    subSequence(lineStart, lineEnd).toString();
-            if (line.contains(IMPORT) && !line.contains("*")) {
-                imports.add(line.replace(IMPORT, "").replace(";", "").trim());
+        PsiFile psiFile;
+        if (psiInstance != null && (psiInstance.getPsiFile(document)) != null) {
+            psiFile = psiInstance.getPsiFile(document);
+            if (psiFile != null) {
+                PsiElement[] psiRootChildren = psiFile.getChildren();
+                PsiElement[] importListChildren;
```
Comment (ScrapCodes) [inline plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/EditorDocOps.java]: There is no need to declare these variables outside the loop. You can declare and assign at same place.

File: plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/EditorDocOps.java [hunk 7]
```diff
@@ -85,17 +93,31 @@
...
+                PsiElement[] importsElementList;
```
Comment (ScrapCodes) [inline plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/EditorDocOps.java]: same for this.

File: plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/EditorDocOps.java [hunk 8]
```diff
@@ -85,17 +93,31 @@
...
+                    for (PsiElement element : psiRootChildren) {
+                        if (element.getNode().getElementType().toString().equals(IMPORT_LIST)) {
+                            importListChildren = element.getChildren();
+                            for (PsiElement importElement : importListChildren) {
+                                if (importElement.getNode().getElementType().
+                                        toString().equals(IMPORT_STATEMENT)) {
+                                    importsElementList = importElement.getChildren();
+                                    if (importsElementList[2] != null) {
```
Comment (ScrapCodes) [inline plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/EditorDocOps.java]: Why are you selecting the `[2]` second index everytime, can you add comment explaining if it is unavoidable ?

File: plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/EditorDocOps.java [hunk 9]
```diff
@@ -159,4 +182,44 @@ private void getCompletePackage(final PsiPackage completePackage,
...
+    }
+
+    private Set<String> getExcludeImports(final String excludeImport) {
+        Set<String> excludeImports = new HashSet<String>();
+
+        if (!excludeImport.isEmpty()) {
+            String[] excludeImportsArray = excludeImport.split(",");
+
+            for (String imports : excludeImportsArray) {
+                  String replacingDot = imports.replace(".", "\\.");
```
Comment (ScrapCodes) [inline plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/EditorDocOps.java]: Why do you need to do this replacement ?

Comment (ScrapCodes) [inline plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/EditorDocOps.java]: I suppose this http://docs.oracle.com/javase/7/docs/api/java/util/regex/Matcher.html#quoteReplacement(java.lang.String) does the same thing in more idiomatic way.

File: plugins/idea/pluginTests/src/test/scala/com/imaginea/betterdocs/ExtractImportSuite.scala [hunk 10]
```diff
@@ -19,10 +19,12 @@ package com.imaginea.betterdocs
 
 import java.io.InputStream
 
+import com.intellij.openapi.command.impl.DummyProject
 import com.intellij.openapi.editor.Document
 import com.intellij.openapi.editor.impl.DocumentImpl
+import com.intellij.openapi.project.Project
 import org.apache.commons.io.IOUtils
-import org.scalatest.{BeforeAndAfterAll, FunSuite}
+import org.scalatest.{Ignore, BeforeAndAfterAll, FunSuite}
```
Comment (ScrapCodes) [inline plugins/idea/pluginTests/src/test/scala/com/imaginea/betterdocs/ExtractImportSuite.scala]: I don't think this import is necessary.

File: plugins/idea/pluginTests/src/test/scala/com/imaginea/betterdocs/ExtractImportSuite.scala [hunk 11]
```diff
@@ -34,8 +36,9 @@ class ExtractImportSuite extends FunSuite with BeforeAndAfterAll {
 
   private val document: Document = new DocumentImpl(fileContents)
 
-  test("Extracted imports should match the imports in java file.") {
-    val editorDocOps = new EditorDocOps().getImports(document)
+  ignore("Extracted imports should match the imports in java file.") {
```
Comment (ScrapCodes) [inline plugins/idea/pluginTests/src/test/scala/com/imaginea/betterdocs/ExtractImportSuite.scala]: Can you add comment here explaining why we have ignored this test ?

**Comment by ScrapCodes** [comment]
@amit-shar I have added a few comments around, fix them soon. It would be great to merge this quick. + It is good if we sync our code formatter  somehow.

File: plugins/idea/pluginTests/src/test/scala/com/imaginea/betterdocs/ExtractImportSuite.scala (hunk 11 above)
Comment (amit-shar) [inline plugins/idea/pluginTests/src/test/scala/com/imaginea/betterdocs/ExtractImportSuite.scala]: As getImports method has been changed and it is taking Project object as a parameter. Currently we are not getting Project object while executing tests. We need to explore this thing to get the Project objects. (//to-do)

Comment (ScrapCodes) [inline plugins/idea/pluginTests/src/test/scala/com/imaginea/betterdocs/ExtractImportSuite.scala]: Add this comment in the code. And also add TODO: explore mocking.

File: plugins/idea/pluginTests/src/test/scala/com/imaginea/betterdocs/ExtractImportSuite.scala (hunk 10 above)
Comment (amit-shar) [inline plugins/idea/pluginTests/src/test/scala/com/imaginea/betterdocs/ExtractImportSuite.scala]: Review feedback

File: plugins/idea/pluginTests/src/test/scala/com/imaginea/betterdocs/ExtractImportSuite.scala (hunk 11 above)
Comment (amit-shar) [inline plugins/idea/pluginTests/src/test/scala/com/imaginea/betterdocs/ExtractImportSuite.scala]: Added. Review deedback

File: plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/SettingsPanel.java (hunk 4 above)
Comment (amit-shar) [inline plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/SettingsPanel.java]: Fixed indentation.

File: plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/EditorDocOps.java (hunk 6 above)
Comment (amit-shar) [inline plugins/idea/betterdocsidea/src/main/java/com/imaginea/betterdocs/EditorDocOps.java]: Review feedback

**Pull request closed without merging**